[RFC] get_keymap API

tjdevries opened the pull request:
Discussion in issue #6123 . Will put more info once features have been decided on.

Two new api functions:

`nvim_get_keymap(mode)`

and

`nvim_buf_get_keymap(buffer, mode)`

tjdevries committed:
Add api function get keymap

nvim_get_keymap(mode)
nvim_buf_get_keymap(buffer, mode)

justinmk commented:
re https://github.com/neovim/neovim/issues/6123#issuecomment-285068783

> What about instead of passing in a string, you could pass in a dictionary.
> This could also let us add more options in the future.

This is something I've been meaning to nail down, so we can extend API functions in a backwards-compatible way. Per msgpack RPC protocol, function parameters are sent as an array "over the wire". But in most clients, those parameters are bound to an arity. E.g., I just tried adding a `String dummy` parameter to `nvim_get_api_info`, and the python client broke.

How about this: design the API with reasonable parameters, but add a dummy dict as the last parameter for future expansion. E.g. in the case of `get_keymap`:

```c
nvim_get_keymap(String mode, String scope, Dictionary reserved_, Error *err)
```

I'll document this pattern in `:help develop.txt` somewhere.

cc @bfredl

justinmk commented:
replying to @ZyX-I comment https://github.com/neovim/neovim/issues/6123#issuecomment-285091028

> I do not like the idea of dummy parameters: how would you know in advance that it is (not) needed? 

The contract of such the "reserved" Dictionary is that any args added there in the future _must_ be optional.

> Better versioned API somewhere, was not this discussed already?

We do version the API. But if we need to extend a function we have to create a new function, we can't add optional arguments.

> When initiating connection request a specific version and when altering API keep an adapter for some specified time so that dispatcher may select proper one

That would be better, but that's not implemented. And it's more complicated. Let's not let perfect get in the way of good.

tjdevries commented:
I'm not sure I like scope being a string. 

I think scope should be a dictionary and should have values like what I mentioned in my other comment:

```vim
{'global': false, 'buffer': false}  " Returns empty list
{'global': false, 'buffer': true}   " Returns list of only buffer mappings
{'global': true, 'buffer': false}   " Returns list of only global mappings.
{'global': true, 'buffer': true}    " Returns list of mappings, with possible overlapping mappings (or we could try and eliminate those)
```

This will make it a lot more clear about what we're doing and make it so we don't have to do any string parsing in the API call. Would that be acceptable?

tjdevries commented:
The other items, leaving a dummy dict and adding an error seem fine with me. Seems like a good in-between solution until we have a more formal way to define adapters for different versions.

ZyX-I commented:
> The contract of the "reserved" Dictionary is that any args added there in the future must be optional.

I did not mean this. I mean that there is no such dictionary in existing API functions and we do not know whether such dictionary will (not) be needed in the future. I do not like the idea of attaching such dictionary to each and every function, this looks ugly. Just as well we may have an agreement that *all* arguments are *only* keyword arguments and never pass any argument lists in first place, dispatcher should know default values and which values are required (either through some syntax extension or just “this type always have X as the default”).

> That would be better, but that's not implemented. And it's more complicated. Let's not let perfect get in the way of good.

Reserved dictionaries do not sound like anything “good”. They look like something not functioning, like dummy functions throwing NotImplementedError if called: you always add a dictionary, but you do not even know whether they will ever be allowed to be non-empty. And you can’t afford not throwing an error for dictionary with unknown arguments: who knows how significant that unknown argument was.

> This will make it a lot more clear about what we're doing and make it so we don't have to do any string parsing in the API call. Would that be acceptable?

“String parsing” is easier then traversing a Dictionary.

justinmk commented:
```vim
{'global': false, 'buffer': false}  " Returns empty list
{'global': false, 'buffer': true}   " Returns list of only buffer mappings
{'global': true, 'buffer': false}   " Returns list of only global mappings.
{'global': true, 'buffer': true}    " Returns list of mappings, with possible overlapping mappings (or we could try and eliminate those)
```

Why do we need the last one? And we definitely don't need the first one. The last one is not different than calling `get_keymap` twice, then merging the results. And it can be done in a single RPC call with `nvim_call_atomic`. So a string parameter avoids an unnecessarily complicated client UX.

justinmk commented:
> I do not like the idea of attaching such dictionary to each and every function, this looks ugly.

I mostly agree, but it's _more_ ugly to add new functions like `nvim_foo2` with slightly different arguments. And when we get the version adapters the ugly "reserved" Dictionary functions can be removed by that very mechanism.

ZyX-I commented:
This would be isolated in a source code, possibly even in a separate file, not something always visible to everyone.

justinmk commented:
On second thought, I have great trepidation about that "version adapter" thing. Anyone who's maintained multiple versions of e.g. a REST API knows that maintaining multiple versions of an API is very difficult. And having the "version adapter" will make us want to use it, and over-use it, and navigating the code will be unpleasant. 

We should follow KISS approach for the time being.

Adding a Dictionary called `optional` isn't that ugly.

jamessan commented:
> After more thought, I have great trepidation about that "version adapter" thing. Anyone who's maintained multiple versions of e.g. a REST API knows that maintaining multiple versions of an API is very difficult, complexity increases non-linearly.

That's why you just add on to the API, instead of removing things.  In many cases, the old API's implementation should be able to just call the new API, so there's not much old code to maintain.  Does it really matter if there's `nvim_foo` and `nvim_foo2`?  Either way (dictionary or new function), the user needs to read documentation to see what's available in which API version.

The incrementing number pattern is used quite successfully in Subversion (e.g. [docs] about the svn_client_checkout family of functions and their [deprecated] implementations).

If there's already a reason to have a dictionary argument, then sure go with that.  If it's just being added as a placeholder, then it seems that just adding a new function later is a better option.

[docs]: https://subversion.apache.org/docs/api/1.9/group__clnt__wc__checkout.html
[deprecated]: https://svn.apache.org/viewvc/subversion/tags/1.9.5/subversion/libsvn_client/deprecated.c?revision=1770702&view=markup#l2512

justinmk commented:
> That's why you just add on to the API, instead of removing things.

No objection there, and that's the thrust of the "API level" concept that we already follow. And the proposed `optional` Dictionary is for [adding](https://github.com/neovim/neovim/pull/6236#issuecomment-285095996) parameters only.

> If there's already a reason to have a dictionary argument, then sure go with that. If it's just being added as a placeholder, then it seems that just adding a new function later is a better option.

Ok. But the (third) approach that I was reacting to is the ["version adapter"](https://github.com/neovim/neovim/issues/6123#issuecomment-285091028) idea, which could be used to _remove_ and re-order parameters. It's a nice idea but could lead us into a quagmire.

justinmk commented:
> If there's already a reason to have a dictionary argument, then sure go with that. If it's just being added as a placeholder, then it seems that just adding a new function later is a better option.

The nice thing about `optional` is that it signals (to devs and client users) that the changes, if any, are:

1. optional
2. accretive

With numbered functions this signal is lost, the docs must be inspected to determine all dimensions of change.

ZyX-I commented:
There is fourth approach: replace argument list with argument dictionary, make dispatcher handle it. Allows default values for all arguments if implemented properly\*, adding new arguments, reordering them. But adds some overhead on processing dictionary.

\* Should be relatively easy to add, including into the current dispatcher, though, of course, only for last N arguments. I think I could do anything required easily, except for writing lpeg grammar without general refactoring of the existing grammar: they managed to produce a thing which has almost nothing in common with anything else I know, not readable and with lots of visual noise. For possible “general refactoring” output check `gendeclarations.lua`.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,65 @@ Dictionary nvim_get_color_map(void)
+
+        Error err = ERROR_INIT;
+
+        maparg_dict = nvim_call_function(cstr_to_string("maparg"), args, &err);
```
For string literals there is STATIC_CSTR_TO_STRING.

tjdevries replied:
Cool! Didn't know about that one.

ZyX-I commented on src/nvim/getchar.c:
```diff
@@ -4261,3 +4261,21 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
+  bool buffer_maps = (STRNCMP(scope, "buffer", 6) == 0);
+  bool global_maps = (STRNCMP(scope, "global", 6) == 0);
+
+  if (buffer_maps) {
```
STRNCMP should be here, and I think global maps should be first. Do not do two comparisons where one is enough.

ZyX-I replied:
Though why `STRNCMP`? This makes arguments like `globalXXX` valid. Should be `strcmp`: no macros (macros was created by Bram only because he uses `char_u`, this function uses `char`), no `n`.

tjdevries replied:
Hadn't thought that through all the way. Didn't know the macros were Vim items.

ZyX-I commented on src/nvim/getchar.c:
```diff
@@ -4261,3 +4261,21 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
   return str[i] == NUL;  // matched the whole string
 }
+
+mapblock_T * get_maphash(int index, char *scope)
```
`const char *const`. And FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE. Also no space after asterisk.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,65 @@ Dictionary nvim_get_color_map(void)
+        maparg_dict = nvim_call_function(cstr_to_string("maparg"), args, &err);
+
+        if (err.set) {
+          ADD(mappings, STRING_OBJ(cstr_to_string("this one didn't work")));
```
I would use `assert(false)` here: missing description for an existing map means that something is broken, better raise the possibility it will be reported.

Though I do not like the idea of using `maparg` function in first place: you just need to copy and refactor ~15 mostly trivial lines from `get_maparg`, this is not enough to bother with reusing code.

tjdevries replied:
I plan on abstracting out a few of the things from maparg to be a separate function. I want them to use the same thing for when we add descriptions or potentially other items.

I just haven't gotten around to it. I pushed to the PR so that we could have a discussion about the API level items, I'm still working on a lot of the implementation and testing.

Thanks for the comments thus far though!

tjdevries replied:
Refactored the function out so that it could be used in both places without all the extra checking.

brcolow commented on test/functional/api/map_spec.lua:
```diff
@@ -0,0 +1,99 @@
+      nvim('command', 'nnoremap foo bar')
+      -- Only one mapping available
+      -- Should be the same as the dictionary we supplied earlier
+      -- and the dicationary you would get from maparg
```
typo: dicationary => dictionary

tjdevries replied:
Fixed... unless I am blind again ;)

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,69 @@ Dictionary nvim_get_color_map(void)
+///
+/// @param mode
+/// @param scope
+Array nvim_get_keymap(String mode, String scope)
```
When rebasing on latest master `FUNC_API_SINCE(2)` will be needed.

tjdevries replied:
Added.

bfredl commented on test/functional/api/map_spec.lua:
```diff
@@ -0,0 +1,99 @@
+  describe('original maparg', function()
+    it('should return a dictionary', function()
+      nvim('command', 'nnoremap foo bar')
+      eq('bar', nvim('call_function', 'maparg', {'foo'}))
```
can use `funcs.maparg('foo')`

tjdevries replied:
Thanks.

tjdevries commented:
I updated the first comment for a description, but here's the status:

New api function: nvim_get_keyamp(String mode, String scope)

Returns a list of mappings for the mode ('n', 'i', ...) or scope ('buffer', 'global')

I have not implemented the reserved or error additions, since I wasn't sure if that's going to be the case of how we're doing things.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +695,56 @@ Dictionary nvim_get_color_map(void)
+    while (current_maphash) {
+      // Check for correct mode
+      if ((int_mode & current_maphash->m_mode) != 0) {
+        typval_T *dict = xmalloc(sizeof(typval_T));
```
You don’t neet to allocate `typval_T` with xmalloc, all functions which deal with `typval_T*` normally expect that you pass something allocated *on the stack* because most of time this is exactly what is being done. But you actually need no typval_T: I would suggest to 

1. Allocate (`dict_T *const dict = dict_alloc();`) and free (`dict_free(dict);`) out of the cycle.
2. Reuse allocated dictionary by using `dict_clear(dict)` in the cycle.
3. Since all functions, but `vim_to_object` expect `dict_T*` you may construct `typval_T` right in the function call: `vim_to_object((typval_T[]) { { .v_type = VAR_DICT, .vval.v_dict = dict } })` if I am not mistaking.

Also I see that you allocate `dict`, but I do not see you freeing it, not to mention unreferencing the dictionary. There should be a memory leak.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +695,56 @@ Dictionary nvim_get_color_map(void)
+      continue;
+    }
+
+    while (current_maphash) {
```
Should not it be

```C
for (mapblock_T *current_maphash = get_maphash(i, scope.data); current_maphash; current_maphash = current_maphash->m_next)
```

? And you definitely need no `if()` with `continue` at all, even with your code: while cycle just will not be executed. Also `current_maphash` variable declaration must be in the cycle, not at the start of the function.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +695,56 @@ Dictionary nvim_get_color_map(void)
+  }
+
+  int i = 0;
+  for (i = 0; i < MAX_MAPHASH; i++) {
```
Remove the above line, `for (int i = 0`.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +695,56 @@ Dictionary nvim_get_color_map(void)
+  // Convert the string mode to the integer mode
+  // that is stored within each mapblock
+  int int_mode = get_map_mode((char_u **)(&(mode.data)), 0);
+  int buffer_local = 0;
```
This is boolean: `const bool buffer_local = strcmp(scope.data, "buffer") == 0;`.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -13358,26 +13358,31 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
   }
 }
 
+void mapblock_fill_dict(dict_T *dict, mapblock_T *mp, int buffer_local)
```
FUNC_ATTR_NONNULL_ALL and `buffer_local` is a clear boolean.

ZyX-I replied:
Additionally I would suggest using `dict_T *const` and `const mapblock_T *const`, especially the second (state that you are not going to modify the `mapblock_T`, `const` after `*` in this case is largerly irrelevant because it only states that variable will not be reassigned inside a function).

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -13358,26 +13358,31 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+  dict_add_nr_str(dict, "expr",    mp->m_expr    ? 1L : 0L, NULL);
+  dict_add_nr_str(dict, "silent",  mp->m_silent  ? 1L : 0L, NULL);
+  dict_add_nr_str(dict, "sid",     (long)mp->m_script_ID, NULL);
+  dict_add_nr_str(dict, "buffer",  (long)buffer_local, NULL);
```
@justinmk How about making that a buffer number or zero?

ZyX-I replied:
Note that this would be backwards-incompatible. Fine when checking `if mp.buffer`, not fine when checking actual value. Help explicitly states that `"buffer"` is 1 for a buffer local mapping.

ZyX-I replied:
Probably worth doing this for `get_keymap`, leaving `maparg()` as-is.

justinmk replied:
:+1:

ZyX-I commented:
@justinmk Also some API refactoring: *remove* scope argument from `nvim_get_keymap` (function will always return global mappings), add `nvim_buf_get_keymap` (obviously, always returns buffer mappings, but for the specified buffer and not for the current one).

ZyX-I commented:
I think that scope argument should be inconsistent with the API: if you need something *entity*-local (buffer-, window-, etc) you go to `nvim_buf` API with specific argument, never have some argument like `scope`. Currently there are only APIs for variables and options with this property (BTW, missing `set_current_dir`, that is only global), let’s continue.

Additionally the workflow “change to buffer, get keymap, change back” is not something you want to do constantly: too many side-effects when buffers are changed.

justinmk commented:
Sounds good

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+//      else: The desired buffer ID
+//
+// @returns An array of maparg() like dictionaries describing mappings
+Array nvim_buf_get_keymap(String mode, Integer buffer_id)
```
This should be in `api/buffer.c`, also first argument should be `Buffer buffer`.

ZyX-I replied:
Use `ArrayOf(Dictionary)` for return type.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+// Get a lit of dictionaries describing buffer-local mappings
+// @param mode The abbreviation for the mode
+// @param buffer_id
+//      0: Current buffer
```
this behavior is shared between all `nvim_buf` functions and need not be mentioned in the docstring.

bfredl commented on src/nvim/eval.c:
```diff
@@ -13358,24 +13358,56 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+//      else: The desired buffer ID
+//          Assumed to be a valid buffer ID
+//  @param compatible True for compatible with old maparg() dict
+void mapblock_fill_dict(dict_T *const dict, const mapblock_T *const mp, long buffer_id, bool compatible)
```
why not simply send in the whatever value `"buffer"` should be set to?  buffer_id is not used otherwise inside this function. `get_maparg` can just continue to send in `buffer_local` and if the `nvim_buf` function does `find_buffer_by_handle` directly like the others, the `buffer==0` logic can stay there and need not be replicated.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+//      0: Current buffer
+//      else: The desired buffer ID
+//
+// @returns An array of maparg() like dictionaries describing mappings
```
Docstring comments start with `///`, *not* `//`. Doxygen won’t pick up `//`.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
 }
 
+// Helper function for nvim_get_keymap and nvim_buf_get_keymap
+// @param mode The abbreviation for the mode
```
Missing empty line after summary. Also I would suggest two spaces before and after parameter name.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
 
+// Helper function for nvim_get_keymap and nvim_buf_get_keymap
+// @param mode The abbreviation for the mode
+// @param buffer_id
```
This should be `buf_T*` and either NULL or pointer to the buffer. Check API functions in `src/nvim/api/buffer.c`: they all start with getting `buf_T*`.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -13358,24 +13358,56 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+//      0: Current buffer
+//      else: The desired buffer ID
+//          Assumed to be a valid buffer ID
+//  @param compatible True for compatible with old maparg() dict
```
This is not a docstring comment. Additionally two spaces before `@param`, should be one.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -13358,24 +13358,56 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
-      dict_add_nr_str(dict, "sid",     (long)mp->m_script_ID, NULL);
-      dict_add_nr_str(dict, "buffer",  (long)buffer_local, NULL);
-      dict_add_nr_str(dict, "nowait",  mp->m_nowait  ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "mode",    0L, (char_u *)mapmode);
```
Missing summary.

ZyX-I commented on src/nvim/getchar.c:
```diff
@@ -4220,3 +4220,28 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
   return str[i] == NUL;  // matched the whole string
 }
+
+// @param index The index in the maphash[]
```
Missing summary. And this is not a docstring comment.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+  return mappings;
+}
+
+// Get a list of dictionaries describing global (i.e. non-buffer) mappings
```
Empty line after summary. Not a docstring.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+}
+
+
+// Get a lit of dictionaries describing buffer-local mappings
```
Empty line after summary. Not a docstring.

ZyX-I replied:
Typo: `lit`, should be `list`.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+// Get a list of dictionaries describing global (i.e. non-buffer) mappings
+// @param mode The abbreviation for the mode
+// @returns An array of maparg() like dictionaries describing mappings
+Array nvim_get_keymap(String mode)
```
Use `ArrayOf(Dictionary)` for return type.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+
+  for (int i = 0; i < MAX_MAPHASH; i++) {
+    for (mapblock_T *current_maphash = get_maphash(i, (long)buffer_id);
+        current_maphash;
```
Too dedented: should be aligned with `current_maphash`. Would also suggest using `const mapblock_T *`.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+        current_maphash = current_maphash->m_next) {
+      // Check for correct mode
+      if ((int_mode & current_maphash->m_mode) != 0) {
+        mapblock_fill_dict(
```
```C
mapblock_fill_dict(dict, current_maphash, buf, false);
ADD(mappings, vim_to_object(
    (typval_T[]) { { .v_type = VAR_DICT, .vval.v_dict = dict } }));
```

first fits on one line just fine (especially since you are supposed to use `buf_T*` which never needs any cast); current variant of `ADD` call could not pass the linter, so I suggested something more condensed which should pass.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -695,6 +696,73 @@ Dictionary nvim_get_color_map(void)
+//      else: The desired buffer ID
+//
+// @returns An array of maparg() like dictionaries describing mappings
+static Array get_keymap_helper(String mode, Integer buffer_id)
```
Avoid naming anything "helper". All functions are helpful, so it doesn't add information. 

Just `keymap_array()` is an OK name for this, analogous to `cursor_shape_dict()`.

tjdevries commented:
There are now two functions:

`nvim_get_keymap(mode)`

and

`nvim_buf_get_keymap(buffer, mode)`

ZyX-I commented on src/nvim/api/buffer.c:
```diff
@@ -446,6 +446,27 @@ Integer nvim_buf_get_changedtick(Buffer buffer, Error *err)
+
+  if (!buf) {
+    Array rv = ARRAY_DICT_INIT;
+    return rv;
```
Just return (Array)ARRAY_DICT_INIT.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -13358,26 +13358,38 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+
+  dict_add_nr_str(dict, "lhs",     0L, lhs);
+  dict_add_nr_str(dict, "rhs",     0L, mp->m_orig_str);
+  dict_add_nr_str(dict, "noremap", mp->m_noremap ? 1L : 0L, NULL);
```
By the way, one thing I missed: if !compatible it should have 2 when `<script>` was used. I.e. `compatible ? !!mp->m_noremap : (mp->m_noremap == REMAP_SCRIPT ? 2 : !!mp->m_noremap)`. Currently `<script>` is not distinguishable with `noremap`, but it is essential if you want to correctly restore a mapping.

ZyX-I commented on test/functional/api/map_spec.lua:
```diff
@@ -0,0 +1,115 @@
+  local insert_table = local_copy(foo_bar_map_table)
+  insert_table['mode'] = 'i'
+
+  describe('original maparg', function()
```
Original maparg should be checked in test/functional/eval, not here.

ZyX-I commented on test/functional/api/map_spec.lua:
```diff
@@ -0,0 +1,115 @@
+  -- The table will be the same except for the mode
+  local foo_bar_insert_string = 'inoremap foo bar'
+  local insert_table = local_copy(foo_bar_map_table)
+  insert_table['mode'] = 'i'
```
Missing mapping with `<script>`.

ZyX-I replied:
And `<expr>`. And, especially, mapping with special characters like `<SID>` (need `helpers.source`) and `<F12>`.

ZyX-I commented on test/functional/api/map_spec.lua:
```diff
@@ -0,0 +1,115 @@
+      nvim('command', 'nnoremap not_going to_check')
+      nvim('command', foo_bar_insert_string)
+
+
```
Too many blank lines.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -13358,26 +13358,38 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+  dict_add_nr_str(dict, "buffer",  buffer_value, NULL);
+
+  dict_add_nr_str(dict, "nowait",  mp->m_nowait  ? 1L : 0L, NULL);
+  dict_add_nr_str(dict, "mode",    0L, (char_u *)mapmode);
```
And you need to rebase, `dict_add_nr_str` function is no longer there. Now have `tv_dict_add…` functions, separate for strings and numbers.

ZyX-I commented on src/nvim/getchar.c:
```diff
@@ -99,11 +101,10 @@ static int block_redo = FALSE;
+// Each mapping is put in one of the MAX_MAPHASH hash lists,
+// to speed up finding it.
+static mapblock_T *(maphash[MAX_MAPHASH]);
+static int maphash_valid = false;
```
May make it a boolean, should not break anything as far as I see.

ZyX-I commented on src/nvim/api/private/helpers.c:
```diff
@@ -971,3 +972,47 @@ static void set_option_value_err(char *key,
+/// @param  mode  The abbreviation for the mode
+/// @param  buffer_id  Buffer handle. -1 For global map array
+/// @returns An array of maparg() like dictionaries describing mappings
+ArrayOf(Dictionary) keymap_array(String mode, Integer buffer_id)
```
This should accept `buf_T*` (NULL or buf). It is not an API function, you don’t need to restrict types.

ZyX-I commented on src/nvim/getchar.c:
```diff
@@ -4220,3 +4221,27 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
+///
+/// @param  index  The index in the maphash[]
+/// @param  buffer_id  -1 for global, else a particular buffer
+mapblock_T *get_maphash(int index, long buffer_id)
```
`buf_T*`

ZyX-I commented on src/nvim/getchar.c:
```diff
@@ -4220,3 +4221,27 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
+  if (buffer_id < 0) {
+    return maphash[index];
+  } else {
+    buf_T *buf = handle_get_buffer((handle_T)(int)buffer_id);
```
This cast is incorrect, you could just cast to `handle_T`. Though this call should not be there in any case.

ZyX-I commented:
C code now LGTM.

bfredl commented on src/nvim/eval.c:
```diff
@@ -12064,26 +12064,54 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+    noremap_value = mp->m_noremap ? 1L : 0L;
+
+    // Force 0 or 1 for buffer_value
+    buffer_value = !!buffer_value;
```
is this needed? seems `check_map` already only returns 0 or 1

tjdevries replied:
Currently, it is redundant, but I figured it would be best to be explicit about what the values will be. I can remove it if you think that's best.

bfredl replied:
Having this check here is rather strange, why would the caller of `mapblock_fill_dict` ever be unable to provide the correct `buffer_value` directly? The `local_ptr` return of `check_map` is only used for the value of the "buffer", so there is no reason it would take some other value.

tjdevries replied:
That makes sense. I agree.

justinmk commented on src/nvim/api/private/helpers.c:
```diff
@@ -965,3 +966,47 @@ static void set_option_value_err(char *key,
+    buffer_value = 0;
+  } else {
+    buffer_value = buf->handle;
+  }
```
ternary saves 6 lines

justinmk commented on src/nvim/getchar.c:
```diff
@@ -4228,3 +4229,21 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
+    return maphash[index];
+  } else {
+    return buf->b_maphash[index];
+  }
```
ternary

justinmk commented on test/functional/api/map_spec.lua:
```diff
@@ -0,0 +1,99 @@
+
```
consider naming the file `keymap_spec`, to parallel the API function names

justinmk commented on test/functional/eval/map_functions_spec.lua:
```diff
@@ -0,0 +1,81 @@
+local source = helpers.source
+
+describe('maparg and api', function()
+  before_each(clear)
```
describe('mappings', function()

ZyX-I replied:
`describe('maparg()'`

ZyX-I replied:
No, `describe('maparg()'`. And remove unneeded nesting level.

justinmk commented on test/functional/eval/map_functions_spec.lua:
```diff
@@ -0,0 +1,81 @@
+      noremap=1,
+    }
+
+  describe('original maparg', function()
```
describe('maparg()', function()

ZyX-I replied:
Replied in changes request.

justinmk commented on test/functional/eval/map_functions_spec.lua:
```diff
@@ -0,0 +1,81 @@
+      eq(foo_bar_map_table, funcs.maparg('foo', 'n', false, true))
+    end)
+
+    it('should return an empty string when no map is present', function()
```
don't need "should" in test descriptions, it's assumed

tjdevries commented:
Okay, I think I addressed the comments that were made. Added about ~100 tests, so I feel pretty confident things are working well. Some test old behavior.

Do I need to add documentation somewhere? Or does it get generated from the api functions?

ZyX-I commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,243 @@
+    eq({insert_table}, nvim('get_keymap', 'i'))
+  end)
+
+  it('works for other modes with modifiers', function()
```
This talks about “modifiers” but actually checks only `<nowait>`. Should be “… with <nowait>”.

tjdevries replied:
This is just a smaller set of the tests I have below, so I removed it. It doesn't add anything.

ZyX-I commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,243 @@
+    global_and_buffer_test(mode_list[mode], 'expr', '<expr>', 1, 1, 0, 0)
+  end
+
+  -- noremap will now be 2 if script was used, which is not backwards compatible
```
There was no API function previously, this is perfectly backwards compatible. Better state “which is different from maparg”.

ZyX-I commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,243 @@
+  global_and_buffer_test('nmap', 'noremap', '<script>', 2, 2, 0, 0)
+  global_and_buffer_test('nnoremap', 'noremap', '<script>', 2, 2, 1, 1)
+
+  -- buffer will return the buffer ID, which is not backwards compatible
```
Same.

ZyX-I commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,243 @@
+    ]])
+    eq(1, eval('g:maparg_test_var'))
+
+    eq(':let g:maparg_test_var = 1<CR>', nvim('get_keymap', 'n')[1]['rhs'])
```
What I asked is what appears in the lhs and rhs when there are special characters like `<F12>`. This only checks `rhs` (`<CR>`), but not `lhs`: better compare the full table returned by `get_keymap`.

ZyX-I commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,243 @@
+    eq(0, eval('g:maparg_test_var'))
+    source([[
+    call feedkeys("\<F12>")
+    ]])
```
There is `[helpers.]feed('<F12>')`, though checking that mapping works is not required. `source` is an overkill unless you do need this: it is working with a temporary file thus is slow and should be omitted when possible.

ZyX-I commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,243 @@
+        buffer_on_result, option, option_token, map, mode), function()
+      make_new_windows(new_windows)
+      nvim('command', map .. ' <buffer> ' .. option_token .. ' foo bar')
+      local result = nvim('buf_get_keymap', 0, mode)[1][option]
```
You are not checking that nvim_buf_get_keymap is working for non-current buffer anywhere.

Additionally I would recommend throwing away `nvim`, there is `[helpers.]*meths`: `meths.get_keymap(mode)`, `curbufmeths.get_keymap(mode)` and `bufmeths.get_keymap(buf, mode)`, `[helpers.]command` for `meths.command`: this variant is more convenient. Should be possible to translate just by executing (in order present)

```Vim
:%s/nvim('buf_\(\w\+\)', 0, /curbufmeths.\1(/g
:%s/nvim('buf_\(\w\+\)', /bufmeths.\1(/g
:%s/nvim('command', /command(/g
:%s/nvim('\(\w\+\)', /meths.\1/g
```

ZyX-I commented:
Documentation is generated.

ZyX-I commented:
Failing tests due to an issue in let_spec:

```
[ RUN      ] ...is/build/neovim/neovim/test/functional/eval/let_spec.lua @ 24: :let command :unlet self-referencing nod
e in a List graph #6070
[       OK ] ...is/build/neovim/neovim/test/functional/eval/let_spec.lua @ 24: :let command :unlet self-referencing nod
e in a List graph #6070 (2.11 ms)
./test/helpers.lua:27: cannot open ./Xtest-tmpdir/nvimUtlurV: No such file or directory

stack traceback:
        ./test/helpers.lua:27: in function 'glob'
        ./test/helpers.lua:195: in function 'check_cores'
        ./test/functional/helpers.lua:628: in function <./test/functional/helpers.lua:626>
```

Interesting, why. And where is that `Xtest-tempdir` coming from, `ag -u` can’t find it.

justinmk commented:
> And where is that Xtest-tempdir coming from, ag -u can’t find it.

@ZyX-I The tests use a workspace-local temp dir since https://github.com/neovim/neovim/pull/6395

tjdevries commented:
I do not experience that failure when I run `make test` locally.

tjdevries commented:
I'd like to add something like this:

```
$ git diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index d0f2b0e..e528be0 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -12093,6 +12093,13 @@ void mapblock_fill_dict(dict_T *const dict,
     // Distinguish between <script> mapping
     // If it's not a <script> mapping, check if it's a noremap
     noremap_value = mp->m_noremap == REMAP_SCRIPT ? 2 : !!mp->m_noremap;
+
+    // Also return the name of the <script>
+    char *file_name = get_script_name((int)mp->m_script_ID);
+    if (file_name != NULL) {
+      tv_dict_add_str(dict, S_LEN("script_name"), (const char*)file_name);
+    }
+    xfree(file_name);
   }

   tv_dict_add_str(dict, S_LEN("lhs"), (const char *)lhs);
diff --git a/src/nvim/ex_cmds2.c b/src/nvim/ex_cmds2.c
index eeace78..e62c74e 100644
--- a/src/nvim/ex_cmds2.c
+++ b/src/nvim/ex_cmds2.c
@@ -2988,6 +2988,15 @@ theend:
 }


+char *get_script_name(int sid)
+{
+  if (SCRIPT_ITEM(sid).sn_name != NULL) {
+    return SCRIPT_ITEM(sid).sn_name;
+  } else {
+    return NULL;
+  }
+}
+
 /// ":scriptnames"
 void ex_scriptnames(exarg_T *eap)
 {
```

Would that be acceptable?

I'd prefer to just add a `nvim_get_script_name(Integer script_id)` because I can't seem to find anywhere to get this that's built-in to nvim.

tjdevries commented:
I guess I could remove the if statement in the `get_script_name` but that's the idea I'm looking for at least.

ZyX-I commented:
@tjdevries This is not the only possible source of script ID. So no, not needed here. It would be nice to add a separate which converts script number to the file name, but in another PR.

Also your code is incorrect:

1. It is useless to check x for NULL just to return x if it is not NULL and NULL if it is. I guess you assumed an allocation given that you use `xfree()`, but
2. What for do you need an allocation at all? `tv_dict_add_str` is copying a string in any case, it is not requiring to transfer ownership to it.
3. And additionally `char *` does not need to be cast to `const char *`. Though `get_script_name` should be returning `const char *` since it does not need to return an allocated string.

ZyX-I commented:
Note: when creating a separate function check for script ID validity. I am also not sure whether you do not need a check here: are SIDs modeline, user input and `--cmd` argument valid?

When answering a question I found that you do *not* need any additional functions, there already is `get_scriptname`.

ZyX-I commented:
And it can even easily be refactored to return `const char *`: there is exactly one user of it.

tjdevries commented:
@ZyX-I I must have searched for the wrong name! Thanks for pointing out that function. I'll make a PR later for adding that to the API and not include this in the current PR.

tjdevries commented:
Well then I think I've addressed all the comments for this PR. Anything else for me to do?

ZyX-I commented on src/nvim/getchar.c:
```diff
@@ -4228,3 +4229,17 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
+/// @param  index  The index in the maphash[]
+/// @param  buf  The buffer to get the maphash from. NULL for global
+mapblock_T *get_maphash(int index, buf_T *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
```
You state that buffer is not NULL here, compiler was nice enough to determine this because I missed.

ZyX-I commented on test/functional/eval/map_functions_spec.lua:
```diff
@@ -0,0 +1,120 @@
+
+  it('works with <expr>', function()
+    source([[
+    let counter = 0
```
Such literals should actually appear indented, it is incorrect to have them on the same level.

ZyX-I commented:
And you need to merge with master or rebase there: not sure, but you may get test failures due to merging in #6489 (specifically, a busted update). I would actually recommend squashing into a single commit and rebasing: most of the commit messages are not very informative.

justinmk commented:
re: busted: I cleared travis cache for this PR.

Quickbuild will fail because a merged PR uses helpers.execute(); I'm fixing that now.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -12064,26 +12064,51 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
   }
 }
 
+/// Fill a dictionary with all applicable maparg() like dictionaries
```
It is strange: GH does not highlight this file starting from this line. All black.

tjdevries commented:
Anything I should be looking to add here?

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -698,6 +698,18 @@ Dictionary nvim_get_color_map(void)
+/// @param  mode  The abbreviation for the mode
+/// @returns  An array of maparg() like dictionaries describing mappings
+ArrayOf(Dictionary) nvim_get_keymap(String mode)
+  FUNC_API_SINCE(2)
```
4 spaces here

bfredl commented on src/nvim/eval.c:
```diff
@@ -12064,26 +12064,51 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+  if (compatible) {
+    // Keep old compatible behavior
+    // This is unable to determine whether a mapping is a <script> mapping
+    noremap_value = mp->m_noremap ? 1L : 0L;
```
In one branch you use `? 1L : 0L` and the other `!!`, why not the same?

bfredl commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,250 @@
+    -- and the dictionary you would get from maparg
+    -- since this is a global map, and not script local
+    eq({foo_bar_map_table}, meths.get_keymap('n'))
+    eq(
```
need not newline directy after `eq(` even if the sides are on different lines.

bfredl commented on test/functional/eval/map_functions_spec.lua:
```diff
@@ -0,0 +1,120 @@
+    nvim('command', 'inoremap <script> hello world')
+    nvim('command', 'inoremap this that')
+    eq(
+      funcs.maparg('hello', 'i', false, true)['noremap'],
```
why not assert the actual values?

tjdevries replied:
I wanted to show that it was equal to the result you'd get from `maparg()`

bfredl commented on test/functional/api/keymap_spec.lua:
```diff
@@ -0,0 +1,250 @@
+    local buffer_table = local_copy(foo_bar_map_table)
+    buffer_table['buffer'] = 1
+
+    command(foo_bar_buffer_string)
```
just move the string literal here

bfredl commented on src/nvim/api/private/helpers.c:
```diff
@@ -965,3 +966,41 @@ static void set_option_value_err(char *key,
+
+  // Convert the string mode to the integer mode
+  // that is stored within each mapblock
+  int int_mode = get_map_mode((char_u **)(&(mode.data)), 0);
```
this would mutate `mode.data`, which is really unexpected, better use temporary `char_u *p = mode.data`.

bfredl commented:
Otherwise LGTM

tjdevries commented:
Fixed comments, and rebased onto master.

tjdevries committed:
Bump API version of keymap functions

tjdevries commented:
Rebased onto master and bumped API version of the two functions made here.

tjdevries commented:
@justinmk I just rebased and pushed again, I think tests should pass. Anything that needs to get done for this to be merged?

tjdevries commented:
got the green check marks :)

bfredl commented on src/nvim/api/buffer.c:
```diff
@@ -453,6 +453,26 @@ Integer nvim_buf_get_changedtick(Buffer buffer, Error *err)
+/// @param[out]  err   Error details, if any
+/// @returns An array of maparg() like dictionaries describing mappings
+ArrayOf(Dictionary) nvim_buf_get_keymap(Buffer buffer, String mode, Error *err)
+  FUNC_API_SINCE(3)
```
4 spaces

bfredl commented:
Otherwise LGTM

tjdevries commented:
I think I got it, had some struggles with rebasing to fixup the indentation commit

justinmk commented:
If I'm reading QB correctly, keymap_spec is resulting in consistent failures since this was merged. Maybe some interaction with the intervening commits.

```
06:00:14,564 INFO  - # test/functional/api/keymap_spec.lua @ 215: get_keymap returns script numbers for global maps
```

tjdevries commented:
Hmm, the only reason I can think that would be is if it returns a different number for the `SID` value there than on my machine. I'm not sure how to get the current `SID` value.

The tests all pass locally for me. How can I debug this?

justinmk commented:
I'm a bit mystified by the QB failure, it is not clear. Perhaps we should create a PR that reverts this one, to see if that makes QB happy, then proceed from there.

ZyX-I commented:
SIDs are incremented, starting from one. To make that robust you just need to call `[helpers.]clear()` so it would restart Neovim and thus restart a counter.

ZyX-I commented:
Which you actually do in before_each().

justinmk commented:
Doesn't look like this PR is the cause of QB hang, rather f4fddbfb778a.

tjdevries commented:
:) Okay. Glad to hear it's working fine then.

oni-link commented on src/nvim/getchar.c:
```diff
@@ -4237,3 +4238,17 @@ static bool typebuf_match_len(const uint8_t *str, int *mlen)
+mapblock_T *get_maphash(int index, buf_T *buf)
+    FUNC_ATTR_PURE
+{
+  if (index > MAX_MAPHASH) {
```
The comparison should be `>=`, otherwise an index of `MAX_MAPHASH (= 256)` could access an element **after**
the array `maphash`.

justinmk replied:
Thanks @oni-link ! Fixed on master.
